node/libs/protobuf_build/src/syntax.rs

```diff
@@ -0,0 +1,244 @@
+        self.0.parent().map(|p| Self(p.into()))
+    }
+
+    /// Derives a proto path from an input path by replacing the $CARGO_MANIFEST_DIR/<input_root> with <proto_root>.
+    pub(super) fn from_input_path(
+        path: &Path,
+        input_root: &InputPath,
+        proto_root: &ProtoPath,
+    ) -> anyhow::Result<ProtoPath> {
+        Ok(ProtoPath(
+            proto_root
+                .0
+                .join(path.strip_prefix(&input_root.abs()?).unwrap()),
+        ))
+    }
+
+    /// Converts ProtoPath to Path.
+    pub(super) fn to_path(&self) -> &Path {
+        &self.0
+    }
+}
+
+impl fmt::Display for ProtoPath {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        self.0.display().fmt(fmt)
+    }
+}
+
+type Part = String;
+
+/// A rust module/type name that the generated code is available at. Although
+/// generated code is location agnostic (it can be embedded in an arbitrary module within the crate),
+/// you need to manually (in the Config) specify the rust modules containing the generated code
+/// of the dependencies, so that it can be referenced from the newly generated code.
+#[derive(Clone, PartialEq, Eq)]
+pub struct RustName(Vec<Part>);
+
+impl RustName {
+    /// Concatenates 2 rust names.
+    pub fn join(mut self, suffix: impl Into<Self>) -> Self {
+        self.0.extend(suffix.into().0);
+        self
+    }
+}
+
+impl fmt::Display for RustName {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str(&self.0.join("::"))
+    }
+}
+
+impl From<&str> for RustName {
+    fn from(s: &str) -> Self {
+        Self(s.split("::").map(Part::from).collect())
+    }
+}
+
+/// A rust module representation.
+/// It is used to collect the generated protobuf code.
+#[derive(Default)]
+pub(super) struct RustModule {
+    /// Nested modules which transitively contain the generated code.
+    modules: BTreeMap<Part, RustModule>,
+    /// Code of the module.
+    code: String,
+}
+
+impl RustModule {
+    /// Returns a reference to a given submodule.
+    pub(crate) fn sub(&mut self, path: &RustName) -> &mut Self {
+        let mut m = self;
+        for part in &path.0 {
+            m = m.modules.entry(part.into()).or_default();
+        }
+        m
+    }
+
+    /// Appends code to the module.
+    pub(crate) fn append(&mut self, code: &str) {
+        self.code += code;
+    }
+
+    fn collect(&self) -> String {
+        let mut entries = vec![self.code.clone()];
+        entries.extend(
+            self.modules
+                .iter()
+                .map(|(name, m)| format!("pub mod {name} {{ {} }}\n", m.collect())),
+        );
+        entries.join("")
+    }
+
+    /// Collects the code of the module and formats it.
+    pub(crate) fn format(&self) -> anyhow::Result<String> {
+        let s = self.collect();
+        Ok(prettyplease::unparse(
+            &syn::parse_str(&s).with_context(|| format!("syn::parse_str({s:?})"))?,
```

Comment:
You can have `RustModule` wrap a `syn::File` internally (in general, dealing with `syn` types rather than strings) in order to get rid of AST -> string -> AST conversions. I'm happy to try doing this in a follow-up PR, so that it the basic functionality is delivered faster.